lib/bolt_server/transport_app.rb

```diff
@@ -142,6 +139,84 @@ def run_task(target, body)
       end
     end
 
+    def run_task(target, body)
+      validate_schema(@schemas["action-run_task"], body)
+
+      task_data = body['task']
+      task = Bolt::Task::PuppetServer.new(task_data['name'], task_data['metadata'], task_data['files'], @file_cache)
+      task_helper(target, task, body['parameters'] || {})
+    end
+
+    def extract_install_task(target)
+      unless target.plugin_hooks['puppet_library']['task']
+        raise BoltServer::RequestError,
+              "Target must have 'task' plugin hook"
+      end
+      install_task = target.plugin_hooks['puppet_library']['task'].split('::', 2)
+      install_task << 'init' if install_task.count == 1
+      install_task
+    end
+
+    def apply_prep(target, body)
+      validate_schema(@schemas["action-apply_prep"], body)
+      plugins_tarball = if (tarball = @file_cache.get_cached_plugin_tarball(body['versioned_project'], 'fact_plugins'))
+                          tarball
+                        else
+                          new_tarball = build_project_plugins_tarball(body['versioned_project']) do |mod|
+                            search_dirs = []
+                            search_dirs << mod.plugins if mod.plugins?
+                            search_dirs << mod.pluginfacts if mod.pluginfacts?
+                            search_dirs
+                          end
+                          @file_cache.cache_plugin_tarball(body['versioned_project'], 'fact_plugins', new_tarball)
+                          new_tarball
+                        end
+      task_datas = in_bolt_project(body['versioned_project']) do |context|
```

Comment:
Hmm, this is a bit concerning because it means we'll be acquiring then releasing a _global_ mutex O(N) times, where N is the # of targets and could be thousands... I don't know how much of a performance hit that may have, if any, but it is probably worth investigating sooner rather than later.

Reply:
If we find that there could be a performance hit, then one option would be to have these endpoints accept the task data in the request. That would mean Orch would need to calculate that info, meaning we'd need to update the apply-* transport methods to include that in the signature.. At that point, it feels like we'd be defeating what is to me the key purpose of doing this in bolt-server -- to avoid cluttering the apply-* transport method's signatures.

Another option would be to cache the task data too, but that may have its own issues if the tasks are from the builtin content since that can change (which begs the Q. if we can assume that a versioned project is truly immutable due to its dependency on the built-in content, which is mutable).

Reply:
Hmm, my point re: caching the task data may be a non-issue if we cache everything on the job ID instead of the versioned project since it's much less likely for e.g. the built-in content to change during a running job than it is during the lifetime of a versioned project.

Given that we already cache something on job ID (e.g. [upload_file](https://github.com/puppetlabs/bolt/blob/main/lib/bolt_server/transport_app.rb#L163)), I kind of lean towards doing that here for all of the apply stuff (plugin tarballs, task datas, etc.). Spiritually, each of those bits of computation represent something happening in a specific _job_ so it makes sense to cache said computation under that job (like upload_file).

Reply:
yeah i thought about this. Couple thoughts. 

1. Project_ref is great for caching. It is unique and immutable. So by caching on that we dont risk dupicating work on jobs run against the same project. 
2. It would be great to remove puppetserver from this equation. This would remove the need to use that mutex at all. I will look in to caching the task metadata because i think that should be easy enough, but i think the better long term goal will be to use file content locally for projects.

Reply:
I guess the task data is a bit different than plugin tarballs though because the plugin tarball and (for now) the custom facts tasks will be the same per-project, but i guess it is possible for different targets to have different agent install tasks. For those i would likely need to cache by job id.

Reply:
actually the combination of project_ref and task name should be unique. Implemented that.